### ISSUE TITLE
Problem restoring files, burp complaining about missing restoreprefix

### ISSUE THREAD

**Issue by deajan** [grke/burp#823]
Hello,

I was trying to restore a backup to a different client via a server initiated restore, without success, so I traced the problem down to this:

Server: CentOS 7 x64 with burp-2.3.6
Client: Windows 10 x64 with burp-2.3.6

On the server, there was a "restore" file created in the directory of the client (via burp-ui), containing the follwing:
```
backup = 44
regex = ^C\:/
restoreprefix = c:\backup
overwrite = 1
orig_client = oldclient.local
```

Running `burp -a t` on the client:
```
C:\Program Files\burp\bin>burp.exe -a t
2019-08-12 16:56:52: burp.exe[11600] windows drives detected: C
2019-08-12 16:56:52: burp.exe[11600] Connecting to burp.server.local:4971
2019-08-12 16:57:13: burp.exe[11600] Could not connect toburp.server.local:4971
2019-08-12 16:57:13: burp.exe[11600] Failing over
2019-08-12 16:57:13: burp.exe[11600] Connecting to burp.server.local:443
2019-08-12 16:57:13: burp.exe[11600] auth ok
2019-08-12 16:57:13: burp.exe[11600] Server version: 2.3.6
2019-08-12 16:57:13: burp.exe[11600] nocsr ok
2019-08-12 16:57:13: burp.exe[11600] SSL is using cipher: DHE-RSA-AES256-GCM-SHA384 TLSv1.2 Kx=DH       Au=RSA  Enc=AESGCM(256) Mac=AEAD

2019-08-12 16:57:13: burp.exe[11600] extra_comms_begin ok:autoupgrade:incexc:orig_client:uname:failover:srestore:sincexc:counters_json:msg:forceproto=1:rshash=blake2:
2019-08-12 16:57:13: burp.exe[11600] Server wants to initiate a restore
2019-08-12 16:57:13: burp.exe[11600] Client accepts.
2019-08-12 16:57:13: burp.exe[11600] The client side must specify a restoreprefix!
2019-08-12 16:57:13: burp.exe[11600] extra comms failed
```

Hmmm... Burp complains about a missing restoreprefix, which seems to be given in the `restore`file. No Idea why the client should specify a restore prefix.
Tried specifying a restore prefix anyway:
```
C:\Program Files\burp\bin>burp.exe -a t -d c:\backup
2019-08-12 17:00:08: burp.exe[5320] windows drives detected: C
2019-08-12 17:00:08: burp.exe[5320] Connecting to burp.fmi-groupe.fr:4971
2019-08-12 17:00:29: burp.exe[5320] Could not connect to burp.fmi-groupe.fr:4971
2019-08-12 17:00:29: burp.exe[5320] Failing over
2019-08-12 17:00:29: burp.exe[5320] Connecting to burp.fmi-groupe.fr:443
2019-08-12 17:00:29: burp.exe[5320] auth ok
2019-08-12 17:00:29: burp.exe[5320] Server version: 2.3.6
2019-08-12 17:00:29: burp.exe[5320] nocsr ok
2019-08-12 17:00:29: burp.exe[5320] SSL is using cipher: DHE-RSA-AES256-GCM-SHA384 TLSv1.2 Kx=DH       Au=RSA  Enc=AESGCM(256) Mac=AEAD

2019-08-12 17:00:29: burp.exe[5320] extra_comms_begin ok:autoupgrade:incexc:orig_client:uname:failover:srestore:sincexc:counters_json:msg:forceproto=1:rshash=blake2:
2019-08-12 17:00:29: burp.exe[5320] Server wants to initiate a restore
2019-08-12 17:00:29: burp.exe[5320] Client accepts.
2019-08-12 17:00:29: burp.exe[5320] windows drives detected: C
2019-08-12 17:00:29: burp.exe[5320] Restore settings:
2019-08-12 17:00:29: burp.exe[5320] orig_client = 'oldclient.local'
2019-08-12 17:00:29: burp.exe[5320] backup = '44'
2019-08-12 17:00:29: burp.exe[5320] overwrite = 1
2019-08-12 17:00:29: burp.exe[5320] strip = 0
2019-08-12 17:00:29: burp.exe[5320] restoreprefix = 'c:\backup/c:\backup'
2019-08-12 17:00:29: burp.exe[5320] regex = '^C\:/'
2019-08-12 17:00:29: burp.exe[5320] main socket 656: expected 'c:orig_client ok', got 'e:0024:running extra comms failed on server'
2019-08-12 17:00:29: burp.exe[5320] Problem requesting orig_client=oldclient.local
2019-08-12 17:00:29: burp.exe[5320] extra comms failed
```

This time, the restore did not succeed because I didn't manage the "superclient" privileges, but that's not the problem I'm seeing.
Looking at the restoreprefix it's logged ```restoreprefix = 'c:\backup/c:\backup'``, which means the restore process took the restore prefix from the `restore` file, and via the client given `-d` option.
Not giving it via `-d` makes restore fail too.
So there might be a problem with the server initiated restore option.

Can make other tests on request.

Best regards.

**Comment by grke** [comment]
Hello,

Your whole problem may be that you are using backslashes in your configurations.
With burp, you should always use forward slashes.
I suggest changing to forward slashes and seeing if the problem goes away.

**Comment by deajan** [comment]
I've reproduced the problem, which also seems to happen with forward slashes:

restore file:
```
backup = 405
regex = ^C\:\/MAILS\/CUSTOM\.DIC$
restoreprefix = c:/testttt
overwrite = 1
```

I've also tried with `restoreprefix = c:\/testttt`, same result.

At the end I tried the following restore file
```
overwrite = 1
```

All the above happen to produce the same output, ie `The client side must specify a restoreprefix!`.

When launching the client with `burp -a t -c c:/testdir`, client logs:
```
could not mkdir c:/toast/c:: Invalid argument
```

Client is burp 2.3.12, server is burp 2.3.12-git.fdbcf74.

**Comment by grke** [comment]
Thanks for coming back with this, I will try to reproduce/fix soon.

**Comment by grke** [comment]
Sorry for my slowness this week, I haven't forgotten.

**Comment by grke** [comment]
Just trying this out now.

The point of needing to specify the restoreprefix on the client side was to provide some kind of security to the client, so that the server could not just overwrite anything on the client. So, it is supposed to concatenate the server side option (if provided) to the client side option.
However, if you really need to restore the files to exactly where the server says, you can make your client side timed command be 'burp -a t -d /'.

Related information from the UPGRADING file (doesn't specifically mention server-initiated restores):
```
2.2.14
------
For security, the burp client will no longer let you restore over symlinks
that point outside of your desired destination directory. Specifying a
'-d <directory>' with your restore parameters is now mandatory. If you wish
to restore everything back to the original locations, you can give '-d /',
which will also work on Windows clients.
```

But anyway, I tried some combinations, because some of tries looked strange:


These work OK - the files get restored to /tmp/xxx:
```
server: restoreprefix=/
client: -d /tmp/xxx

server: restoreprefix=/tmp/xxx
client: -d /
```

This works OK - the files get restored to /tmp/xxx/blah:
```
server: restoreprefix=/blah
client -d /tmp/xxx
```

This works OK - the restore fails:
```
server: restoreprefix=/tmp/xxx
client: nothing specified
```

This does not work OK - the files get restored to /tmp/xxx/tmp/xxx:
```
server: no restoreprefix
client: -d /tmp/xxx
```

So, there is definitely a problem with not specifying an option on the server side.
I will look at fixing that.

**Comment by deajan** [comment]
@grke Thanks for the clear explanation. I'll have some tests with 2.3.14 once I am near my setup again.
@ziirish I think this breaks the default "server initiated restore" feature of burp-ui. Maybe there should be a word on that behavior directly in the burp-ui interface, as the whole function does not work anymore, unless `burp -a t -d /` is used client side, which is probably configured nowhere.

**Comment by ziirish** [comment]
Agreed, I'll track this on burp-ui's issue tracker.

Just to be sure, here is what should be explained:

> client side, the `-d` flag is required
> restore prefix defined burp-ui/server side will be APPENDED to what's defined with `-d`

right?

**Comment by deajan** [comment]
Looks good to me. Maybe adding this ?

> On Windows, the scheduled task created by Burp does not include -d.

**Comment by grke** [comment]
Closing - fixed in 2.3.14.
